File: web/src/lib/view-model/passport-view-model.ts
import type { BatteryPassport, ChartSegment, Cluster, PassportAppMetadata, PassportDocumentLink } from "../../types/passport";

function valueAt(payload: Record<string, unknown>, key: string) {
  return payload[key] == null ? "" : String(payload[key]);
}

function numberAt(payload: Record<string, unknown>, key: string) {
  const value = Number(payload[key]);
  return Number.isFinite(value) ? value : 0;
}

function dateOnly(value: string) {
  return value ? value.slice(0, 10) : "";
}

function fallbackApp(passport: BatteryPassport): PassportAppMetadata {
  const general = passport.aspects.generalProductInformation?.payload ?? {};
  const carbon = passport.aspects.carbonFootprintForBatteries?.payload ?? {};
  const carbonStages = Array.isArray(carbon.carbonFootprintPerLifecycleStage)
    ? (carbon.carbonFootprintPerLifecycleStage as Array<{ lifecycleStage?: string; carbonFootprint?: number }>)
    : [];

  return {
    display: {
      name: valueAt(general, "productIdentifier"),
      modelNumber: valueAt(general, "productIdentifier"),
      serialNumber: valueAt(general, "batteryPassportIdentifier"),
      facilityId: "",
      manufacturerName: "",
    },
    media: {
      batteryImageUrl: "/sample-battery.png",
      batteryImageAlt: "Industrial EV battery pack",
    },
    documents: {
      conformityAssessment: { label: "Conformity assessment", url: "" },
      euDeclarationOfConformity: { label: "EU declaration of conformity ID", url: "" },
      sustainabilityReport: { label: "Sustainability report", url: "" },
      dueDiligenceReport: { label: "Due diligence report", url: "" },
      thirdPartyAudit: { label: "Third party audit", url: "" },
      taxonomyReport: { label: "Taxonomy report", url: "" },
      co2StudyReference: { label: "CO2 study reference", url: valueAt(carbon, "carbonFootprintStudy") },
    },
    charts: {
      materialComposition: [],
      carbonFootprint: carbonStages.map((stage) => ({
        label: String(stage.lifecycleStage ?? ""),
        value: Number(stage.carbonFootprint ?? 0),
        unit: "gCO2e/kWh",
      })),
      recycledContent: [],
    },
    notes: {
      circularity: {
        separateCollection: "",
        wastePrevention: "",
        recycledContentShareVerification: "unverified",
      },
    },
  };
}

function chartTotal(segments: ChartSegment[]) {
  return segments.reduce((total, segment) => total + segment.value, 0);
}

function withPercentages(segments: ChartSegment[]) {
  const total = chartTotal(segments);
  return segments.map((segment) => ({
    ...segment,
    percentage: total > 0 ? Math.round((segment.value / total) * 1000) / 10 : 0,
  }));
}

export function toPassportViewModel(passport: BatteryPassport, options: { clusters?: Cluster[] } = {}) {
  const app = passport.app ?? fallbackApp(passport);
  const general = passport.aspects.generalProductInformation?.payload ?? {};
  const carbon = passport.aspects.carbonFootprintForBatteries?.payload ?? {};
  const performance = passport.aspects.performanceAndDurability?.payload ?? {};
  const batteryCondition = (performance.batteryCondition ?? {}) as Record<string, Record<string, unknown>>;
  const technical = (performance.batteryTechicalProperties ?? {}) as Record<string, unknown>;

  const documents = Object.fromEntries(
    Object.entries(app.documents).map(([key, document]) => [key, document as PassportDocumentLink]),
  ) as PassportAppMetadata["documents"];
  const clusterNames = passport.clusterId
    ? [options.clusters?.find((cluster) => cluster.clusterId === passport.clusterId)?.name ?? passport.clusterId]
    : [];

  return {
    passportId: passport.passportId,
    clusterId: passport.clusterId,
    clusterNames,
    clusterLabel: clusterNames.length ? clusterNames.join(", ") : "No cluster assigned",
    registryId: passport.registryInfo.registryId,
    status: passport.registryInfo.status,
    name: app.display.name,
    modelNumber: app.display.modelNumber || valueAt(general, "productIdentifier"),
    serialNumber: app.display.serialNumber || valueAt(general, "batteryPassportIdentifier"),
    category: valueAt(general, "batteryCategory").toUpperCase(),
    batteryStatus: valueAt(general, "batteryStatus"),
    manufacturedDate: dateOnly(valueAt(general, "manufacturingDate")),
    manufacturerName: app.display.manufacturerName,
    facilityId: app.display.facilityId,
    weight: numberAt(general, "batteryMass"),
    weightLabel: `${numberAt(general, "batteryMass").toFixed(2)}kg`,
    carbonFootprint: valueAt(carbon, "batteryCarbonFootprint"),
    carbonFootprintLabel: `${Number(valueAt(carbon, "batteryCarbonFootprint") || 0).toFixed(2)}gCO2e/kWh`,
    performanceClass: valueAt(carbon, "carbonFootprintPerformanceClass"),
    isValid: passport.validation.isValid,
    verificationState: passport.validation.isValid ? "verified" : "unverified",
    batteryImageUrl: app.media.batteryImageUrl,
    batteryImageAlt: app.media.batteryImageAlt,
    documents,
    materialComposition: {
      total: chartTotal(app.charts.materialComposition),
      segments: withPercentages(app.charts.materialComposition),
    },
    carbonFootprintStages: {
      total: chartTotal(app.charts.carbonFootprint),
      segments: withPercentages(app.charts.carbonFootprint),
    },
    recycledContent: app.charts.recycledContent,
    circularity: app.notes.circularity,
    performance: {
      ratedEnergy: Number(technical.ratedEnergy ?? 0),
      ratedCapacity: Number(technical.ratedCapacity ?? 0),
      ratedMaximumPower: Number(technical.ratedMaximumPower ?? 0),
      nominalVoltage: Number(technical.nominalVoltage ?? 0),
      expectedLifetime: Number(technical.expectedLifetime ?? 0),
      expectedNumberOfCycles: Number(technical.expectedNumberOfCycles ?? 0),
      stateOfCharge: Number(batteryCondition.stateOfCharge?.stateOfChargeValue ?? 0),
      remainingCapacity: Number(batteryCondition.remainingCapacity?.remainingCapacityValue ?? 0),
      remainingEnergy: Number(batteryCondition.remainingEnergy?.remainingEnergyValue ?? 0),
      cycles: Number(batteryCondition.numberOfFullCycles?.numberOfFullCyclesValue ?? 0),
    },
    sections: ["General", "Material composition", "Performance", "Compliance", "Supply chain", "Circularity", "Carbon Footprint"],
  };
}

File: web/Models/ViewModels/PassportViewModel.cs
namespace BatteryPassWeb.Models.ViewModels;

public sealed class PassportViewModel
{
    public string PassportId { get; init; } = string.Empty;
    public string DisplayName { get; init; } = string.Empty;
    public string ModelNumber { get; init; } = string.Empty;
    public string SerialNumber { get; init; } = string.Empty;
    public string ManufacturerName { get; init; } = string.Empty;
    public string FacilityId { get; init; } = string.Empty;
    public string RegistryStatus { get; init; } = string.Empty;
    public string ClusterId { get; init; } = string.Empty;
    public string ClusterLabel { get; init; } = "No cluster assigned";
    public string Category { get; init; } = string.Empty;
    public string BatteryStatus { get; init; } = string.Empty;
    public string ManufacturedDate { get; init; } = string.Empty;
    public double Weight { get; init; }
    public string WeightLabel { get; init; } = string.Empty;
    public bool IsValid { get; init; }
    public string VerificationState { get; init; } = "unverified";
    public string TrustState { get; init; } = "unvalidated";
    public bool TrustIsDirty { get; init; }
    public string TrustLastValidatedAt { get; init; } = string.Empty;
    public int TrustBlockingErrorCount { get; init; }
    public int TrustWarningCount { get; init; }
    public string TrustLatestHash { get; init; } = string.Empty;
    public string TrustLatestRevisionId { get; init; } = string.Empty;
    public string TrustLastSignedAt { get; init; } = string.Empty;
    public string TrustIssuer { get; init; } = string.Empty;
    public string TrustVerificationMethod { get; init; } = string.Empty;
    public string TrustProofStatus { get; init; } = "No proof";
    public string TrustVerificationMessage { get; init; } = string.Empty;
    public string BatteryImageUrl { get; init; } = "/images/compact7.png";
    public string BatteryImageAlt { get; init; } = "Industrial EV battery pack";
    public double CarbonFootprint { get; init; }
    public string CarbonFootprintLabel { get; init; } = string.Empty;
    public string PerformanceClass { get; init; } = string.Empty;
    public PassportPerformanceViewModel Performance { get; init; } = new();
    public PassportOperationsViewModel Operations { get; init; } = new();
    public PassportCircularityViewModel Circularity { get; init; } = new();
    public PassportDocumentsViewModel Documents { get; init; } = new();
    public IReadOnlyList<ChartSegmentViewModel> MaterialCompositionSegments { get; init; } = [];
    public IReadOnlyList<ChartSegmentViewModel> CarbonFootprintSegments { get; init; } = [];
    public IReadOnlyList<RecycledContentChartViewModel> RecycledContentCharts { get; init; } = [];
    public IReadOnlyList<BatteryMaterialRowViewModel> BatteryMaterials { get; init; } = [];
    public double MaterialCompositionTotal { get; init; }
    public double SupplyChainIndex { get; init; }
    public IReadOnlyList<string> Sections { get; init; } =
    [
        "General",
        "Material composition",
        "Performance",
        "Compliance",
        "Supply chain",
        "Circularity",
        "Carbon Footprint"
    ];
}

public sealed class PassportPerformanceViewModel
{
    public double RatedEnergy { get; init; }
    public double RatedCapacity { get; init; }
    public double RatedMaximumPower { get; init; }
    public double NominalVoltage { get; init; }
    public double ExpectedLifetime { get; init; }
    public double ExpectedNumberOfCycles { get; init; }
    public double StateOfCharge { get; init; }
    public double RemainingCapacity { get; init; }
    public double RemainingEnergy { get; init; }
    public double Cycles { get; init; }
}

public sealed class PassportOperationsViewModel
{
    public bool IsActive { get; init; } = true;
    public string LocationSiteName { get; init; } = string.Empty;
    public string LocationAddress { get; init; } = string.Empty;
    public string LocationCity { get; init; } = string.Empty;
    public string LocationCountry { get; init; } = string.Empty;
    public double? LocationLatitude { get; init; }
    public double? LocationLongitude { get; init; }
    public string ContactName { get; init; } = string.Empty;
    public string ContactEmail { get; init; } = string.Empty;
    public string ContactPhone { get; init; } = string.Empty;
    public double? CurrentConsumptionKwh { get; init; }
    public double? CurrentChargeLevelPct { get; init; }
    public double? CurrentVoltageV { get; init; }
    public double? CurrentCurrentA { get; init; }
    public string LatestMeasuredAt { get; init; } = string.Empty;
}

public sealed class PassportCircularityViewModel
{
    public string SeparateCollection { get; init; } = string.Empty;
    public string WastePrevention { get; init; } = string.Empty;
    public string RecycledContentShareVerification { get; init; } = "unverified";
}

public sealed class PassportDocumentsViewModel
{
    public PassportDocumentLinkViewModel ConformityAssessment { get; init; } = new("Conformity assessment", string.Empty);
    public PassportDocumentLinkViewModel EuDeclarationOfConformity { get; init; } = new("EU declaration of conformity ID", string.Empty);
    public PassportDocumentLinkViewModel SustainabilityReport { get; init; } = new("Sustainability report", string.Empty);
    public PassportDocumentLinkViewModel DueDiligenceReport { get; init; } = new("Due diligence report", string.Empty);
    public PassportDocumentLinkViewModel ThirdPartyAudit { get; init; } = new("Third party audit", string.Empty);
    public PassportDocumentLinkViewModel TaxonomyReport { get; init; } = new("Taxonomy report", string.Empty);
    public PassportDocumentLinkViewModel Co2StudyReference { get; init; } = new("CO2 study reference", string.Empty);
}

public sealed record PassportDocumentLinkViewModel(string Label, string Url, string FileId = "", string ContentType = "", string Visibility = "private");

public sealed class ChartSegmentViewModel
{
    public string Label { get; init; } = string.Empty;
    public double Value { get; init; }
    public string Unit { get; init; } = string.Empty;
    public string Color { get; init; } = string.Empty;
    public double Percentage { get; init; }
}

public sealed class RecycledContentChartViewModel
{
    public string Material { get; init; } = string.Empty;
    public double PreConsumerShare { get; init; }
    public double PostConsumerShare { get; init; }
    public double PrimaryMaterialShare { get; init; }
}

public sealed class BatteryMaterialRowViewModel
{
    public string MaterialName { get; init; } = string.Empty;
    public double MaterialMass { get; init; }
    public string ComponentName { get; init; } = string.Empty;
    public bool IsCriticalRawMaterial { get; init; }
}
